lib/parsers/asm-parser-vc.ts

```diff
@@ -296,17 +315,20 @@ export class VcAsmParser extends AsmParser {
                 //   - two compiler generated functions
                 // order by name
                 if (f1.initialLine === f2.initialLine) {
+                    assert(f1.name !== undefined && f2.name !== undefined);
                     return collator.compare(f1.name, f2.name);
```

Comment:
Not 100% sure about the asserts here, in the JS code it wouldn't be a problem to pass non-string input to Intl.Collator.compare

Reply:
I'm pretty unhappy with all the weird asserts in honesty. TS was angry saying that undefined was not allowed here:
```TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.   Type 'undefined' is not assignable to type 'string'.```

Reply:
At leaSt in the `es5` thing my IDE makes up:
```
    interface Collator {
        compare(x: string, y: string): number;
        resolvedOptions(): ResolvedCollatorOptions;
    }
```

so could be IDE only?

Reply:
not just the IDE; also complained with
```
TSError: ⨯ Unable to compile TypeScript:
lib/parsers/asm-parser-vc.ts:319:45 - error TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

319                     return collator.compare(f1.name, f2.name);
                                                ~~~~~~~
```
in the build

Reply:
Maybe `return collator.compare(f1.name ?? "", f2.name ?? "");` could be an option?

Reply:
In any case, this function can't throw, we've had issues before with this comparison

Reply:
Done: at least... I cahnged the collator compare. There are still other `assert`s here, as the alternative would be an exception anyway (I think?) Are you saying this function _shouldn't throw_ partouf (as opposed to it can't throw/)

Reply:
It's a bit of a weighing game, in this 'compare' line, I would say, let it silently continue. If there are other asserts that you think we should know about when they fail, then by all means keep them.

Reply:
This line's fine - it's the other ones that assert non-nullity etc (which I think would "just" have thrown a JS error before if they happened). I'll merge once I get things working (after another think...) and you can always point me back at this and say "i told you so".

Reply:
Argh we can't use `??` 
```
/home/runner/work/compiler-explorer/compiler-explorer/lib/parsers/asm-parser-vc.ts:47
if(f1.initialLine===f2.initialLine){cov_29xlecyj9b().b[52][0]++;cov_29xlecyj9b().s[145]++;return collator.compare((cov_29xlecyj9b().b[53][0]++,f1.name)??(cov_29xlecyj9b().b[53][1]++,''),(cov_29xlecyj9b().b[54][0]++,f2.name)??(cov_29xlecyj9b().b[54][1]++,''));}else{cov_29xlecyj9b().b[52][1]++;cov_29xlecyj9b().s[146]++;assert((cov_29xlecyj9b().b[55][0]++,f1.initialLine!==undefined)&&(cov_29xlecyj9b().b[55][1]++,f2.initialLine!==undefined));cov_29xlecyj9b().s[147]++;return f1.initialLine-f2.initialLine;}}else{cov_29xlecyj9b().b[51][1]++;}// else, order by file
                                                                                                                                                        ^

SyntaxError: Invalid or unexpected token
    at Object.replacementCompile (/home/runner/work/compiler-explorer/compiler-explorer/node_modules/append-transform/index.js:60:13)
    at Object.m._compile (/home/runner/work/compiler-explorer/compiler-explorer/node_modules/ts-node/src/index.ts:1618:[23](https://github.com/compiler-explorer/compiler-explorer/actions/runs/4134952334/jobs/7146752626#step:5:24))```

Reply:
aka:

```
~/d/c/compiler-explorer (mg/tsify-vc|✔) $ make test
/opt/compiler-explorer/node/bin/npm run test

> compiler-explorer@0.0.3 test
> mocha -b -r ts-node/register 'test/**/*.ts' 'test/**/*.js'


/home/matthew/dev/ce/compiler-explorer/lib/parsers/asm-parser-vc.ts:273
                    return collator.compare(f1.name ?? '', f2.name ?? '');
```

Reply:
once we finally upgrade everything we can kill esm and then... we can do this :(